arch/arm64/src/common/arm64_boot.c

```diff
@@ -0,0 +1,185 @@
+
+  reg   = read_sysreg(ICC_SRE_EL3);
+  reg   |= (ICC_SRE_ELX_DFB_BIT |   /* Disable FIQ bypass */
+            ICC_SRE_ELX_DIB_BIT |   /* Disable IRQ bypass */
+            ICC_SRE_ELX_SRE_BIT |   /* System register interface is used */
+            ICC_SRE_EL3_EN_BIT);    /* Enables lower Exception level access to
+                                     * ICC_SRE_EL1 */
+  write_sysreg(reg, ICC_SRE_EL3);
+
+  __ISB();
+}
+
+void arm64_boot_el3_get_next_el(uint64_t switch_addr)
+{
+  uint64_t spsr;
+
+  write_sysreg(switch_addr, elr_el3);
+
+  /* Mask the DAIF */
+
+  spsr  = SPSR_DAIF_MASK;
+  spsr  |= SPSR_MODE_EL2T;
+
+  write_sysreg(spsr, spsr_el3);
+}
+
+void arm64_boot_el2_init(void)
+{
+  uint64_t reg;
+
+  reg   = read_sysreg(sctlr_el2);
+  reg   |= (SCTLR_EL2_RES1 |    /* RES1 */
+            SCTLR_I_BIT |       /* Enable i-cache */
+            SCTLR_SA_BIT);      /* Enable SP alignment check */
+  write_sysreg(reg, sctlr_el2);
+
+  reg   = read_sysreg(hcr_el2);
+  reg   |= HCR_RW_BIT;      /* EL1 Execution state is AArch64 */
+  write_sysreg(reg, hcr_el2);
+
+  reg   = 0U;                   /* RES0 */
+  reg   |= CPTR_EL2_RES1;       /* RES1 */
+  reg   &= ~(CPTR_TFP_BIT |     /* Do not trap SVE, SIMD and FP */
+             CPTR_TCPAC_BIT);   /* Do not trap CPACR_EL1 accesses */
+  write_sysreg(reg, cptr_el2);
+
+  /* Enable EL1 access to timers */
+
+  reg   = read_sysreg(cnthctl_el2);
+  reg   |= (CNTHCTL_EL2_EL1PCEN_EN | CNTHCTL_EL2_EL1PCTEN_EN);
+  write_sysreg(reg, cnthctl_el2);
+
+  zero_sysreg(cntvoff_el2);       /* Set 64-bit virtual timer offset to 0 */
+
+#ifdef CONFIG_ARCH_ARMV8R
+  zero_sysreg(cnthps_ctl_el2);
+#else
+  zero_sysreg(cnthp_ctl_el2);
+#endif
+
+  /* Enable this if/when we use the hypervisor timer.
+   * write_cnthp_cval_el2(~(uint64_t)0);
+   */
+
+  __ISB();
+}
+
+void arm64_boot_el1_init(void)
+{
+  uint64_t reg;
+
+  /* Setup vector table */
+
+  write_sysreg((uint64_t)_vector_table, vbar_el1);
+  __ISB();
+
+  reg   = 0U;                       /* RES0 */
+  reg   |= CPACR_EL1_FPEN_NOTRAP;   /* Do not trap NEON/SIMD/FP initially */
+
+  /* TODO: CONFIG_FLOAT_*_FORBIDDEN */
+
+  write_sysreg(reg, cpacr_el1);
+
+  reg   = read_sysreg(sctlr_el1);
+  reg   |= (SCTLR_EL1_RES1 |    /* RES1 */
+            SCTLR_I_BIT |       /* Enable i-cache */
+            SCTLR_SA_BIT);      /* Enable SP alignment check */
+  write_sysreg(reg, sctlr_el1);
+
+  write_sysreg((~(uint64_t)0), cntv_cval_el0);
+
+  /* Enable these if/when we use the corresponding timers.
+   * write_cntp_cval_el0(~(uint64_t)0);
+   * write_cntps_cval_el1(~(uint64_t)0);
+   */
+
+  __ISB();
+}
+
+/* These simple memset alternatives are necessary
+ * as the function at libc is depend on the MMU
+ * to be active.
+ */
+
+static void boot_early_memset(void *dst, int c, size_t n)
+{
+  uint8_t *d = dst;
+
+  while (n--)
+    {
+      *d++ = c;
+    }
+}
+
+void arm64_boot_primary_c_routine(void)
+{
+  boot_early_memset(_sbss, 0, _ebss - _sbss);
```

Comment:
Can we clear with 64bits loop?

Reply:
No, at this stage, the MMU is not enable, and maybe alignment check is enable
so it's safe operation to clear as  byte

Reply:
Ok. Usually BSS section should be machine word aligned, but I'm fine with what we are having now.
The other platforms use
```
#  define _START_BSS   &_sbss
#  define _END_BSS     &_ebss
```
and
```
  /* Clear .bss.  We'll do this inline (vs. calling memset) just to be
   * certain that there are no issues with the state of global variables.
   */

  for (dest = _START_BSS; dest < _END_BSS; )
    {
      *dest++ = 0;
    }
```
in `__start`. Or do BSS clear in assembly code like `arm_data_initialize:` in `arch/arm/src/armv7-a/arm_head.S`.
This is optional just to keep similar style.

I'm fine to keep code as is and just pointing possible variants

Reply:
add define _START_BSS and _END_BSS
in C level,  we cannot make sure the operation will be alignment in any case or optimization level. So operating by byte is a very conservative but safety approach. 
assembly code can make the guarantee. Actually at the early single core version, I clear BSS in arm64_head.S. but for SMP,  the operation of clear BSS can only be done by primary core, it's a little complicated to clear BSS at arm64_head.S for SMP case, so I move the operation to C level
if have any need for booting performance requirement, maybe it's a optimizing point to move clear BSS to assembly code. i maybe rework it at future